src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1050,67 +1050,69 @@ private bool IsValidFixedVariableInitializer(TypeSymbol declType, SourceLocalSym
-                            // OK
-                            break;
-                        case BoundKind.Conversion:
-                            // The following assertion would not be correct because there might be an implicit conversion after (above) the explicit one.
-                            //Debug.Assert(((BoundConversion)initializerOpt).ExplicitCastInCode, "The assignment conversion hasn't been applied yet, so this must be from source.");
-
-                            // NOTE: Dev10 specifically doesn't report this error for the array or string cases.
-                            Error(diagnostics, ErrorCode.ERR_BadCastInFixed, initializerSyntax);
+                            elementType = ((BoundAddressOfOperator)initializerOpt).Operand.Type;
                             break;
                         case BoundKind.FieldAccess:
                             var fa = (BoundFieldAccess)initializerOpt;
                             if (!fa.FieldSymbol.IsFixed)
                             {
                                 Error(diagnostics, ErrorCode.ERR_FixedNotNeeded, initializerSyntax);
+                                return true;
+                            }
+                            else
+                            {
+                                elementType = ((PointerTypeSymbol)fa.Type).PointedAtType;
                             }
                             break;
+                        case BoundKind.Conversion:
+                            // The following assertion would not be correct because there might be an implicit conversion after (above) the explicit one.
+                            //Debug.Assert(((BoundConversion)initializerOpt).ExplicitCastInCode, "The assignment conversion hasn't been applied yet, so this must be from source.");
+
+                            // NOTE: Dev10 specifically doesn't report this error for the array or string cases.
+                            Error(diagnostics, ErrorCode.ERR_BadCastInFixed, initializerSyntax);
+                            return true;
                         default:
                             // CONSIDER: this is a very confusing error message, but it's what Dev10 reports.
                             Error(diagnostics, ErrorCode.ERR_FixedNotNeeded, initializerSyntax);
-                            break;
+                            return true;
                     }
                 }
+                else
+                {
+                    // convert to generate any additional conversion errors
+                    initializerOpt = GenerateConversionForAssignment(declType, initializerOpt, diagnostics);
+                    return true;
+                }
 
-                initializerOpt = GenerateConversionForAssignment(declType, initializerOpt, diagnostics);
+                initializerOpt = GetFixedLocalCollectionInitializer(initializerOpt, elementType, declType, hasErrors, diagnostics);
```

Comment:
GetSpecialType that takes diagnostics bag does not return null. It would report an error (into the bag) and return an error symbol. 
I think we are fine here and the same possibility existed before the change.

---
In reply to: [125139886](https://github.com/dotnet/roslyn/pull/20548#discussion_r125139886) [](ancestors = 125139886)